### PR TITLE
TrackTfClassifier with Batch evaluation of the DNN

### PR DESCRIPTION
**Pull request by leonardogiannini**
#### PR description:

Switching TrackTfClassifier from single track to batch (size 16) evaluation. The batch size is configurable and can be set to 1.

Batch eval was found to be faster for the cand DNN in mkFit (see [TK POG update 25 April](https://indico.cern.ch/event/1152032/contributions/4841692/attachments/2430728/4162233/mkFit%20-%20update-1.pdf) )

No physics change is expected, but the evaluation is about 3 times faster.

#### PR validation:

selection step speed up going from blue (old) to red (new) -> reducing a bit the overall tracking time
image

the physics is the same: 
see e.g. http://uaf-10.t2.ucsd.edu/~legianni/batchDNNSel-PR/plots_ootb/effandfakePtEtaPhi.pdf
or  http://uaf-10.t2.ucsd.edu/~legianni/batchDNNSel-PR/plots_highPurity/effandfakePtEtaPhi.pdf
for a validation comparison in TTbar with PU.

### CODE REVIEW

**Comment by mmusich** [comment]
please test

**Comment by jpata** [comment]
enable profiling

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> enable profiling

based on https://github.com/cms-sw/cmssw/issues/37816#issuecomment-1131371863
I suspect that this will not work
> Many [ib-run-profiling](https://cmssdt.cern.ch/jenkins/job/ib-run-profiling/6531/console) jobs for el8_amd64_gcc10 are failing

**Comment by jpata** [comment]
@cmsbuild please cancel

File: RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc [hunk 1]
```diff
@@ -34,72 +36,115 @@ namespace {
...
+
+      tensorflow::Tensor input1(tensorflow::DT_FLOAT, {bsize_, 29});
+      tensorflow::Tensor input2(tensorflow::DT_FLOAT, {bsize_, 1});
+
+      for (auto nb = 0; nb < nbatches + 1; nb++) {
+        for (auto nt = 0; nt < bsize_; nt++) {
+          int itrack = nt + bsize_ * nb;
+          if (itrack >= size_in)
+            continue;
+          const auto& trk = tracks.at(itrack);
```
Comment (jpata) [inline RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc:56]: Usually we ask for the no-exception version, especially in performance-critical code.
```suggestion
          const auto& trk = tracks[itrack];
```

File: RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc [hunk 2]
```diff
@@ -34,72 +36,115 @@ namespace {
...
   };
+}  // namespace
+
+template <>
+void trackMVAClassifierImpl::ComputeMVA<void>::operator()(::TfDnn const& mva,
+                                                          reco::TrackCollection const& tracks,
+                                                          reco::BeamSpot const& beamSpot,
+                                                          reco::VertexCollection const& vertices,
+                                                          TrackMVAClassifierBase::MVAPairCollection& mvas) {
+  std::vector<float> scores = mva(tracks, beamSpot, vertices);
```
Comment (jpata) [inline RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc:135]:
```suggestion
  const auto& scores = mva(tracks, beamSpot, vertices);
```
perhaps?

**Review by jpata — COMMENTED** [review]
I had just minor comments.

File: RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc (hunk 1 above)
Comment (leonardogiannini) [reply RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc]: I should have done this already as in 37878. sorry for the double review.

File: RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc (hunk 2 above)
Comment (leonardogiannini) [reply RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc]: same as above

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> **Additional Tests**: PROFILING

In the Run3 workflow the DNN evaluation cost for tracks is down by about a factor of 3, from [6.2 perf counts](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_5_X_2022-05-20-1100/el8_amd64_gcc10/profiling/11834.21/igprofCPU_step3/160) to [2.2 perf counts](https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_12_5_X_2022-05-20-1100/el8_amd64_gcc10/profiling/11834.21/PR-6bcf7a/24880/igprofCPU_step3/486)

This is overall around 1.1% reduction in the reco time.

**Comment by jpata** [comment]
Note that the DQM picks up a number of small differences in tracking like the following:

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-05-20-1100+6bcf7a/50423/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25/Tracking__TrackBuilding_highPtTripletStep_quickAssociatorByHits_mva_fake_mva1_eta.png

Tracking__TrackBuilding_highPtTripletStep_quickAssociatorByHits_mva_fake_mva1_eta

Since the rare DNN irreproducibility was fixed, should we conclude that batching can occasionally result in small changes in the DNN output?

**Comment by slava77** [comment]
> 
> Since the rare DNN irreproducibility was fixed, should we conclude that batching can occasionally result in small changes in the DNN output?

it looks like the differences are at the numerical precision level
I picked the histograms corresponding to the plot and printed bin contents for `fabs(d)>1e-6`
```
bin    IB                  PR           d=IB-PR
19 -0.104025 -0.104027 2.58287e-06
24 0.0390263 0.0390287 -2.42392e-06
26 0.47047 0.470469 1.19209e-06
32 -0.720131 -0.720134 2.98023e-06
38 -0.370042 -0.370039 -3.33786e-06
```

**Comment by jpata** [comment]
+reconstruction
- improves track DNN evaluation time by  batching
- technical (apart from the numerical precision described above)

**Comment by jpata** [comment]
@cms-sw/tracking-pog-l2 et al, should we consider this for a backport to 12_4? The physics changes are negligible, but the timing improvement is significant.

**Comment by mmusich** [comment]
> et al, should we consider this for a backport to 12_4?

It would be indeed desirable.

**Comment by leonardogiannini** [comment]
ok, I'll do the backport to 12_4 later today

**Comment by qliphy** [comment]
+1

**Pull request merged**